File: kh1_src/tabs/inventory.py
from dash import Dash, html, dcc, callback, Input, Output, State, ALL
import kh1_src.kh1_utils as utils


@callback(
    Output("InventoryDiv", "children"),
    Input("InventoryTabs", "value"),
)
def __create_inventory(tab):
    kh1 = utils.kh1
    inventory = None
    if tab == "Consumables":
        interval = lambda x: x in range(0x01, 0x09) or x in range(0x8E, 0x91) or x in range(0x98, 0x9B)
    if tab == "Synthesis Materials":
        interval = lambda x: x in range(0x09, 0x11) or x in range(0x9B, 0x9E) or x > 0xE8
    if tab == "Accessories":
        interval = lambda x: x in kh1.accessory_dict.values() and x > 0
    if tab == "Weapons":
        interval = lambda x: x in kh1.weapon_dict.values() and x > 0
    if tab == "Key Items":
        interval = lambda x: x in range(0x95, 0x98) or x in range(0x9E, 0xE8)
    inventory = html.Div([
        html.Div([
            html.Label(k + ": "),
            dcc.Input(
                id={"type": "Inventory", "index": v},
                type="number",
                value=kh1.inventory[v],
                min=0,
                max=99,
                step=1,
                style={"width": 50},
            ),
        ],
            style={"margin-top": 20, "gap": 10},
        ) for k, v in kh1.item_dict.items()\
        if ("Unused" not in k and interval(v)\
        if tab != "Unused" else "Unused" in k)
    ])
    return inventory

def create_inventory():
    kh1 = utils.kh1
    itabs = dcc.Tabs(id="InventoryTabs", value="Consumables")
    itabs.children = [
        dcc.Tab(label="Consumables", value="Consumables"),
        dcc.Tab(label="Synthesis Materials", value="Synthesis Materials"),
        dcc.Tab(label="Accessories", value="Accessories"),
        dcc.Tab(label="Weapons", value="Weapons"),
        dcc.Tab(label="Key Items", value="Key Items"),
        dcc.Tab(label="Unused", value="Unused"),
    ]
    return html.Div([
        itabs,
        html.Div(id="InventoryDiv"),
    ])

@callback(
    Input({"type": "Inventory", "index": ALL}, "value"),
    State({"type": "Inventory", "index": ALL}, "id"),
)
def inventory_callback(values, ids):
    kh1 = utils.kh1
    try:
        i = 0
        for id in ids:
            idx = id["index"]
            kh1.inventory[idx] = values[i]
            i += 1
    except:
        pass

File: kh1_src/tabs/inventory.js
export function create_inventory() {
    const kh1div = document.getElementById("kh1div");
    kh1div.innerHTML = `
    <div>
        <div>
            <h3>Category:</h3>
            <select id="kh1invtabs">
                <option value="Consumables">Consumables</option>
                <option value="Synthesis Materials">Synthesis Materials</option>
                <option value="Accessories">Accessories</option>
                <option value="Weapons">Weapons</option>
                <option value="Key Items">Key Items</option>
                <option value="Unused">Unused</option>
            </select>
        </div>
        <div id="kh1invdiv"></div>
    </div>`;
    const kh1invtabs = document.getElementById("kh1invtabs");
    function tab_sel() {
        __create_inventory(kh1invtabs.value);
    }
    kh1invtabs.addEventListener("change", tab_sel);
    __create_inventory("Consumables");
}

function __create_inventory(tab) {
    const kh1invdiv = document.getElementById("kh1invdiv");
    let items = null;
    switch (tab) {
        case "Consumables": {
            items = Object.fromEntries(Object.entries(window.kh1.item_dict)
                .filter(([label, value]) => {
                    return !label.includes("Unused") && (
                        (value >= 0x01 && value < 0x09) ||
                        (value >= 0x8E && value < 0x91) ||
                        (value >= 0x98 && value < 0x9B)
                    );
                }));
            break;
        }
        case "Synthesis Materials": {
            items = Object.fromEntries(Object.entries(window.kh1.item_dict)
                .filter(([label, value]) => {
                    return !label.includes("Unused") && (
                        (value >= 0x09 && value < 0x11) ||
                        (value >= 0x9B && value < 0x9E) ||
                        (value >= 0xE9 && value < 0x100)
                    );
                }));
            break;
        }
        case "Accessories": {
            items = Object.fromEntries(Object.entries(window.kh1.accessory_dict)
                .filter(([label, value]) => {
                    return !label.includes("Unused") && value > 0;
                }));
            break;
        }
        case "Weapons": {
            items = Object.fromEntries(Object.entries(window.kh1.weapon_dict)
                .filter(([label, value]) => {
                    return !label.includes("Unused") && value > 0;
                }));
            break;
        }
        case "Key Items": {
            items = Object.fromEntries(Object.entries(window.kh1.item_dict)
                .filter(([label, value]) => {
                    return !label.includes("Unused") && (
                        (value >= 0x95 && value < 0x98) ||
                        (value >= 0x9E && value < 0xE8)
                    );
                }));
            break;
        }
        case "Unused": {
            items = Object.fromEntries(Object.entries(window.kh1.item_dict)
                .filter(([label, value]) => {
                    return label.includes("Unused");
                })
                .sort());
            break;
        }
    };
    const inventory = Object.entries(items)
        .map(([label, value]) => `
        <label style="display: flex; gap: 10px; align-items: center">
            ${label}: 
            <input
                type="number"
                min=0
                max=99
                step=1
                value=${window.kh1.inventory[value]}
                name=${value}
            >
        </label>`)
        .join("");
    kh1invdiv.innerHTML = `
    <div id="inventory">
        <h3>Inventory:</h3>
        ${inventory}
    </div>`;
    inventory_callbacks();
}

function inventory_callbacks() {
    const inventory = document.getElementById("inventory");
    inventory.addEventListener("change", (e) => {
        if (e.target.validity.valid)
            window.kh1.inventory[e.target.name] = e.target.value;
        e.target.value = window.kh1.inventory[e.target.name];
    });
}
